### PR TITLE
Aztec: Show Media picker in the input view

### PR DESCRIPTION
**Pull request by SergioEstevao**
This PR makes the media picker to be displayed on the input view (keyboard) instead of going full screen. 

screen shot 2017-06-21 at 16 04 37

To test:
 - Start the app
 - Enable Aztec
 - Start a new post
 - Tap on add media
 - Select media and tap on done
 - Do the same but press cancel
 - Test tapping on the Grid icon

Needs review: @diegoreymendez, @frosty , @iamthomasbishop

### CODE REVIEW

**Comment by iamthomasbishop** [comment]
@SergioEstevao That looks cool to me, but I'm not super familiar with the media updates that have been in progress so I might defer to @mattmiklic. Does this look right, MT?

**Comment by diegoreymendez** [comment]
@SergioEstevao - Looks good to me, but we should probably feature flag this for now.  I'd like this to avoid publishing this as part of the open beta, to give us time to test it, run it through @iamthomasbishop, and synch with @0nko for the Android side of it.

Let's try to keep the discussion here so all the interested parties can chime in.

Thanks!

**Comment by SergioEstevao** [comment]
@iamthomasbishop @mattmiklic , I think we are now in a funny situation with the media picker, for one side it could be considered work on the Aztec/Editor toolbar and from another it could be considered work on the media updates :).

I do think this interaction model makes thing quicker specially on the editor, as you don't need to jump to another screen.

I also talked with @frosty and for now I'm going to hide this behind a feature flag.

**Comment by SergioEstevao** [comment]
BTW I did target this for 8.0 not 7.9. Is the open beta starting on 7.9 or 8.0 release of the app?

**Comment by diegoreymendez** [comment]
> BTW I did target this for 8.0 not 7.9. Is the open beta starting on 7.9 or 8.0 release of the app?

The open beta starts with v8.0, I believe (since 7.9 is currently coming out).

That said: I still think this should be feature flagged and run through design + Android synch, then closed beta, and open beta.

There's a greater risk in coupling the open beta of Aztec with the new media picker that we should avoid.

**Comment by mattmiklic** [comment]
Agreed with the suggestion to put this behind a feature flag for the time being. Since I'm covering design on the media project for the time being, I'm going to look at this and give feedback on how to align the design with what's been done on Android.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 1]
```diff
@@ -1419,16 +1421,72 @@ extension AztecPostViewController : Aztec.FormatBarDelegate {
         insertAction.isEnabled = !urlFieldText.isEmpty
     }
 
-    func presentMediaPicker(animated: Bool = true) {
+    var mediaInputToolbar: UIToolbar {
+        let toolbar = UIToolbar(frame: CGRect(x: 0, y: 0, width: view.frame.width, height: 44.0))
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: Replace the magic number with a constant.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 2]
```diff
@@ -1419,16 +1421,72 @@ extension AztecPostViewController : Aztec.FormatBarDelegate {
         insertAction.isEnabled = !urlFieldText.isEmpty
     }
 
-    func presentMediaPicker(animated: Bool = true) {
+    var mediaInputToolbar: UIToolbar {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: Readability: Add some spacing in this method to improve readability.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 3]
```diff
@@ -1419,16 +1421,72 @@ extension AztecPostViewController : Aztec.FormatBarDelegate {
...
+        ]
+        for item in toolbar.items! {
+            item.tintColor = WPStyleGuide.aztecFormatBarActiveColor
+            item.setTitleTextAttributes([NSForegroundColorAttributeName: WPStyleGuide.aztecFormatBarActiveColor], for: .normal)
+        }
+        return toolbar
+
+    }
+
+    func mediaAddShowFullScreen(_ sender: UIBarButtonItem) {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: Document.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 4]
```diff
@@ -1419,16 +1421,72 @@ extension AztecPostViewController : Aztec.FormatBarDelegate {
...
+        presentMediaPickerFullScreen(animated: true)
+    }
+
+    func mediaAddInputCancelled(_ sender: UIBarButtonItem) {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: Document.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 5]
```diff
@@ -1419,16 +1421,72 @@ extension AztecPostViewController : Aztec.FormatBarDelegate {
...
+        guard let mediaPicker = mediaPickerInputViewController?.mediaPicker else {
+            return
+        }
+        mediaPickerControllerDidCancel(mediaPicker)
+    }
+
+    func mediaAddInputDone(_ sender: UIBarButtonItem) {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: Document.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 6]
```diff
@@ -1419,16 +1421,72 @@ extension AztecPostViewController : Aztec.FormatBarDelegate {
...
         picker.delegate = self
-        picker.modalPresentationStyle = .currentContext
+        picker.modalPresentationStyle = .currentContext        
+        // Disable the input media picker if we go full screen.
+        mediaPickerInputViewController = nil
+        present(picker, animated: true)
+    }
 
-        present(picker, animated: animated, completion: nil)
+    func presentMediaPicker(fromItem item: FormatBarItem, animated: Bool = true) {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: Readability: Add some spacing in is method to improve readability.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 7]
```diff
@@ -1419,16 +1421,72 @@ extension AztecPostViewController : Aztec.FormatBarDelegate {
...
+    func mediaAddInputDone(_ sender: UIBarButtonItem) {
+        guard let mediaPicker = mediaPickerInputViewController?.mediaPicker,
+              let selectedAssets = mediaPicker.selectedAssets as? [Any]
+        else {
+            return
+        }
+        mediaPickerController(mediaPicker, didFinishPickingAssets: selectedAssets)
+    }
 
+    func presentMediaPickerFullScreen(animated: Bool) {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: This method should probably be private.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift (hunk 2 above)
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: This should probably be private.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift (hunk 6 above)
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: This method should probably be private.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 8]
```diff
@@ -1468,6 +1518,47 @@ extension AztecPostViewController : Aztec.FormatBarDelegate {
         })
     }
 
+    func insertHorizontalRuler() {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: This method should probably be private.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 9]
```diff
@@ -1468,6 +1518,47 @@ extension AztecPostViewController : Aztec.FormatBarDelegate {
...
+        richTextView.replaceRangeWithHorizontalRuler(richTextView.selectedRange)
+    }
+
+    func insertMore() {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: This method should probably be private.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 10]
```diff
@@ -1468,6 +1518,47 @@ extension AztecPostViewController : Aztec.FormatBarDelegate {
...
+        richTextView.replaceRangeWithCommentAttachment(richTextView.selectedRange, text: Constants.moreAttachmentText)
+    }
+
+    func headerLevelForSelectedText() -> Header.HeaderType {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: This method should probably be private.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift (hunk 8 above)
Comment (SergioEstevao) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: This wasn't changed on this PR

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift (hunk 9 above)
Comment (SergioEstevao) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: This wasn't changed on this PR

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift (hunk 10 above)
Comment (SergioEstevao) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: This wasn't changed on this PR

**Comment by SergioEstevao** [comment]
@diegoreymendez ready for another look.

**Comment by SergioEstevao** [comment]
@diegoreymendez ready for another look.

**Comment by SergioEstevao** [comment]
@diegoreymendez ready for another look.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 11]
```diff
@@ -142,7 +142,7 @@ class AztecPostViewController: UIViewController, PostEditor {
     /// Separator View
     ///
     fileprivate(set) lazy var separatorView: UIView = {
-        let v = UIView(frame: CGRect(x: 0, y: 0, width: 44, height: 1))
+        let v = UIView(frame: CGRect(x: 0, y: 0, width: Constants.toolbarHeight, height: 1))
```
Comment (SergioEstevao) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: This was already there, I was just applying the constant but I think this is to draw the bar to separate the plus (media) option from the other items?

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 12]
```diff
@@ -1530,17 +1532,87 @@ extension AztecPostViewController : Aztec.FormatBarDelegate {
...
-    func presentMediaPicker(animated: Bool = true) {
-        trackFormatBarAnalytics(stat: .editorTappedImage)
+    private var mediaInputToolbar: UIToolbar {
+        let toolbar = UIToolbar(frame: CGRect(x: 0, y: 0, width: view.frame.width, height: Constants.toolbarHeight))
+        toolbar.barTintColor = WPStyleGuide.aztecFormatBarBackgroundColor
+        toolbar.tintColor = WPStyleGuide.aztecFormatBarActiveColor
+        toolbar.items = [
+            UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(mediaAddInputCancelled)),
+            UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil),
+            UIBarButtonItem(image: Gridicon.iconOfType(.grid), style: .plain ,target: self, action: #selector(mediaAddShowFullScreen)),
```
Comment (SergioEstevao) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: Good catch.

**Comment by SergioEstevao** [comment]
@frosty I update the code following your comments, please have another look.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 13]
```diff
@@ -1606,6 +1623,15 @@ extension AztecPostViewController : Aztec.FormatBarDelegate {
         let picker = WPInputMediaPickerViewController()
         mediaPickerInputViewController = picker
         richTextView.inputAccessoryView = mediaInputToolbar
+
+        originalLeadingBarButtonGroup = richTextView.inputAssistantItem.leadingBarButtonGroups
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: 👍 It's silly we have to do this!

**Comment by frosty** [comment]
Thanks @SergioEstevao, changes look good 👍 

However, I'm seeing some weird issues when connecting / disconnecting a hardware keyboard. For example:

* Open the editor, and have the hardware keyboard connected.
* Tap the Add Media button to bring up the picker.
* Start typing on the hardware keyboard.
* The media picker disappears, but the media picker toolbar is still visible. The Cancel and Done buttons have no effect.

Also

* With the hardware keyboard disconnected, bring up the picker
* Connect the hardware keyboard.
* Dismiss the picker.
* The suggestions / auto complete toolbar doesn't come back.

**Comment by SergioEstevao** [comment]
@frosty ready for another look.

**Comment by apsist025350** [comment]
เกมเศรษฐี

**Comment by SergioEstevao** [comment]
@frosty ready for another look.

**Pull request merged**